kernel/src/fs/named_pipe.rs

```diff
@@ -1,50 +1,285 @@
+                .inner
+                .reader
+                .state
+                .poll_with(mask, poller, || self.inner.reader.check_io_events()),
+            AccessMode::O_WRONLY => self
+                .inner
+                .writer
+                .state
+                .poll_with(mask, poller, || self.inner.writer.check_io_events()),
+            AccessMode::O_RDWR => {
+                let read_events =
+                    self.inner
+                        .reader
+                        .state
+                        .poll_with(mask, poller.as_deref_mut(), || {
+                            self.inner.reader.check_io_events()
+                        });
+                let write_events = self
+                    .inner
+                    .writer
+                    .state
+                    .poll_with(mask, poller, || self.inner.writer.check_io_events());
+
+                read_events | write_events
+            }
+        }
     }
 }
 
-impl FileLike for NamedPipe {
-    fn read(&self, writer: &mut VmWriter) -> Result<usize> {
-        self.reader.read(writer)
+impl Drop for NamedPipeHandle {
+    fn drop(&mut self) {
+        match self.access_mode {
+            AccessMode::O_RDONLY => {
+                let old_value = self.inner.num_reader.fetch_sub(1, Ordering::Relaxed);
+                if old_value == 1 {
+                    self.inner.reader.state.peer_shutdown();
+                }
+            }
+            AccessMode::O_WRONLY => {
+                let old_value = self.inner.num_writer.fetch_sub(1, Ordering::Relaxed);
+                if old_value == 1 {
+                    self.inner.writer.state.shutdown();
+                }
+            }
+            AccessMode::O_RDWR => {
+                let old_reader_value = self.inner.num_reader.fetch_sub(1, Ordering::Relaxed);
+                if old_reader_value == 1 {
+                    self.inner.reader.state.peer_shutdown();
+                }
+
+                let old_writer_value = self.inner.num_writer.fetch_sub(1, Ordering::Relaxed);
+                if old_writer_value == 1 {
+                    self.inner.writer.state.shutdown();
+                }
+            }
+        }
+    }
+}
+
+impl FileIo for NamedPipeHandle {
+    fn read(&self, writer: &mut VmWriter, status_flags: StatusFlags) -> Result<usize> {
+        if let AccessMode::O_WRONLY = self.access_mode {
+            return Err(Error::with_message(Errno::EBADF, "not opened for reading"));
+        }
+
+        if status_flags.contains(StatusFlags::O_NONBLOCK) {
+            self.try_read(writer)
+        } else {
+            self.wait_events(IoEvents::IN, None, || self.try_read(writer))
+        }
+    }
+
+    fn write(&self, reader: &mut VmReader, status_flags: StatusFlags) -> Result<usize> {
+        if let AccessMode::O_RDONLY = self.access_mode {
+            return Err(Error::with_message(Errno::EBADF, "not opened for writing"));
+        }
```

Comment:
```suggestion
        if !self.access_mode.is_writable() {
            return_errno_with_message!(Errno::EBADF, "the pipe is not opened for writing");
        }
```

I think using `!self.access_mode.is_writable()` is more clearer.

Please update all `read()`/`write()` methods in this file?